airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -0,0 +1,148 @@
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def close_partition(self, partition: Partition) -> None:
+        """
+        Indicate to the cursor that the partition has been successfully processed
+        """
+        raise NotImplementedError()
+
+
+class NoopCursor(Cursor):
+    def observe(self, record: Record) -> None:
+        pass
+
+    def close_partition(self, partition: Partition) -> None:
+        pass
+
+
+class ConcurrentCursor(Cursor):
+    _START_BOUNDARY = 0
+    _END_BOUNDARY = 1
+
+    def __init__(
+        self,
+        stream_name: str,
+        stream_namespace: Optional[str],
+        stream_state: Any,
+        message_repository: MessageRepository,
+        connector_state_manager: ConnectorStateManager,
+        cursor_field: CursorField,
+        slice_boundary_fields: Optional[Tuple[str, str]],
+    ) -> None:
+        self._stream_name = stream_name
+        self._stream_namespace = stream_namespace
+        self._message_repository = message_repository
+        self._connector_state_manager = connector_state_manager
+        self._cursor_field = cursor_field
+        # To see some example where the slice boundaries might not be defined, check https://github.com/airbytehq/airbyte/blob/1ce84d6396e446e1ac2377362446e3fb94509461/airbyte-integrations/connectors/source-stripe/source_stripe/streams.py#L363-L379
+        self._slice_boundary_fields = slice_boundary_fields if slice_boundary_fields else tuple()
+        self._most_recent_record: Optional[Record] = None
+        self._has_closed_at_least_one_slice = False
+
+        # TODO to migrate state. The migration should probably be outside of this class. Impact of not having this:
+        #  * Given a sync that emits no records, the emitted state message will be empty
+        self._state = {  # type: ignore  # waiting for when we implement the state migration
+            "slices": []  # empty for now but should look like `{start: 1, end: 10, parent_id: "id1"}`
+        }
+
+    def observe(self, record: Record) -> None:
+        if self._slice_boundary_fields:
+            # Given that slicing is done using the cursor field, we don't need to observe the record as we assume slices will describe what
+            # has been emitted. Assuming there is a chance that records might not be yet populated for the most recent slice, use a lookback
+            # window
+            return
+
+        if not self._most_recent_record or self._extract_cursor_value(self._most_recent_record) < self._extract_cursor_value(record):
+            self._most_recent_record = record
+
+    def _extract_cursor_value(self, record: Record) -> Comparable:
+        return self._cursor_field.extract_value(record)
+
+    def close_partition(self, partition: Partition) -> None:
+        slice_count_before = len(self._state["slices"])
+        self._add_slice_to_state(partition)
+        if slice_count_before < len(self._state["slices"]):
```

Comment:
would the size of th updated state ever not be greater than the previous count?

Reply:
Yes, if we process a slice without boundaries fields and without any record